polkadot/runtime/parachains/src/assigner_on_demand/tests.rs

```diff
@@ -497,3 +536,127 @@ fn affinity_changes_work() {
 		assert!(OnDemandAssigner::get_affinity_map(para_a).is_none());
 	});
 }
+
+#[test]
+fn new_affinity_for_a_core_must_come_from_free_entries() {
+	// If affinity count for a core was zero before, and is 1 now, then the entry
+	// must have come from free_entries.
+	let parachains =
+		vec![ParaId::from(111), ParaId::from(222), ParaId::from(333), ParaId::from(444)];
+	let core_indices = vec![CoreIndex(0), CoreIndex(1), CoreIndex(2), CoreIndex(3)];
+
+	new_test_ext(GenesisConfigBuilder::default().build()).execute_with(|| {
+		parachains.iter().for_each(|chain| {
+			schedule_blank_para(*chain, ParaKind::Parathread);
+		});
+
+		run_to_block(11, |n| if n == 11 { Some(Default::default()) } else { None });
+
+		// Place orders for all chains.
+		parachains.iter().for_each(|chain| {
+			place_order(*chain);
+		});
+
+		// There are 4 entries in free_entries.
+		let start_free_entries = OnDemandAssigner::get_free_entries().len();
+		assert_eq!(start_free_entries, 4);
+
+		// Pop assignments on all cores.
+		core_indices.iter().enumerate().for_each(|(n, core_index)| {
+			// There is no affinity on the core prior to popping.
+			assert!(OnDemandAssigner::get_affinity_entries(*core_index).is_empty());
+
+			// There's always an order to be popped for each core.
+			let free_entries = OnDemandAssigner::get_free_entries();
+			let next_order = free_entries.peek();
+
+			// There is no affinity on the paraid prior to popping.
+			assert!(OnDemandAssigner::get_affinity_map(next_order.unwrap().para_id).is_none());
+
+			match OnDemandAssigner::pop_assignment_for_core(*core_index) {
+				Some(assignment) => {
+					// The popped assignment came from free entries.
+					assert_eq!(
+						start_free_entries - 1 - n,
+						OnDemandAssigner::get_free_entries().len()
+					);
+					// The popped assignment has the same para id as the next order.
+					assert_eq!(assignment.para_id(), next_order.unwrap().para_id);
+				},
+				None => panic!("Should not happen"),
+			}
+		});
+
+		// All entries have been removed from free_entries.
+		assert!(OnDemandAssigner::get_free_entries().is_empty());
+
+		// All chains have an affinity count of 1.
+		parachains.iter().for_each(|chain| {
+			assert_eq!(OnDemandAssigner::get_affinity_map(*chain).unwrap().count, 1);
+		});
+	});
+}
+
+#[test]
+#[should_panic]
+fn queue_index_ordering_is_unsound_over_max_size() {
+	// NOTE: Unsoundness proof. If the number goes sufficiently over the max_queue_max_size
+	// the overflow will cause an opposite comparison to what would be expected.
+	let mut max_num = u32::MAX - ON_DEMAND_MAX_QUEUE_MAX_SIZE;
+	// 0 < some large number.
+	assert_eq!(QueueIndex(0).cmp(&QueueIndex(max_num + 1)), Ordering::Less);
+}
+
+#[test]
+fn queue_index_ordering_works() {
+	// The largest number accepted in in the queue index.
+	let mut max_num = ON_DEMAND_MAX_QUEUE_MAX_SIZE;
```

Comment:
The index is actually allowed to be larger, just the size of the queue must never exceed this value.